subsys/net/l2_wifi_if_conn/l2_wifi_conn.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(l2_wifi_mgr_conn);
+#include <zephyr/net/net_if.h>
+#include <zephyr/net/wifi_mgmt.h>
+#include <zephyr/net/net_event.h>
+#include <net/wifi_mgmt_ext.h>
+#include <net/l2_wifi_connect.h>
+#include <zephyr/net/conn_mgr_connectivity.h>
+
+#define WIFI_SHELL_MGMT_EVENTS (NET_EVENT_WIFI_CONNECT_RESULT | \
+				NET_EVENT_WIFI_DISCONNECT_RESULT)
+
+static struct net_mgmt_event_callback net_l2_mgmt_cb;
+int connection_status;
+
+
+int net_l2_wifi_connect(struct conn_mgr_conn_binding *const binding);
+int net_l2_wifi_disconnect(struct conn_mgr_conn_binding *const binding);
+
+/* Delayable work used to handle Wi-Fi connection timeouts. */
+static void wifi_conn_work_handler(struct k_work *work);
+K_THREAD_STACK_DEFINE(l2_wifi_wq_stack_area, CONFIG_L2_WIFI_CONN_WQ_STACK_SIZE);
+struct k_work_q wifi_conn_wq;
+static K_WORK_DELAYABLE_DEFINE(wifi_conn_work, wifi_conn_work_handler);
+
+/* Store a local reference to the connection binding */
+struct conn_mgr_conn_binding  *wifi_conn_binding;
+
+static void wifi_conn_work_handler(struct k_work *work)
+{
+
+	if (connection_status != WIFI_STATE_COMPLETED) {
+		net_l2_wifi_disconnect(wifi_conn_binding);
+	}
+}
+
+static int wifi_conn_timeout_schedule(struct conn_mgr_conn_binding *const binding)
+{
+	int rc;
+
+	rc = net_mgmt(net_request_wifi_connect_stored, binding->iface, null, 0);
+
+	if (rc) {
+		log_err("net management connect_stored request failed: %d\n", rc);
+		return -ENOEXEC;
+	}
+
+	if (binding->timeout > 0) {
+		rc = k_work_schedule_for_queue(&wifi_conn_wq, &wifi_conn_work,
+					       K_SECONDS(binding->timeout));
+	}
+
+	return rc;
+}
+
+static void net_l2_wifi_connect_result(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status =
+		(const struct wifi_status *) cb->info;
+
+	if (status->status) {
+		connection_status = WIFI_STATE_DISCONNECTED;
```

Comment:
I think you need to initialize another timeout here if persistence is enabled, right?

Reply:
Here we are relying on supplicant's inherent retry logic to implement persistence. In current implementation, NET_EVENT_WIFI_CONNECT_RESULT is invoked with failure only in case of timeout. So, invoking another timeout is not required here.

Reply:
> Then, an hour later, connection to the AP is lost

This triggers a `NET_EVENT_WIFI_CMD_DISCONNECT_RESULT` and then we restart the workqueue to handle the timeout for this attempt (as WPA supplicant starts the next attempt after this).

> but I had thought @krish2718 mentioned that the built-in timeout will only occur on first connection attempt. If that's true, then I think there is a problem here

I might have misled you here, my entire `timeout` explanation was based on the assumption that `wifi_mgmt_ext` uses `wifi_mgmt` (which invokes the the Zephyr's implementation callbacks for `wifi_mgmt` where we handle the timeout), but `wifi_mgmt_ext` directly talks to WPA supplicant (`wpa_cli` calls), so, this `timeout` mechanism doesn't apply here, apologies, last discussion with @maxd-nordic  was to move to `wifi_mgmt` I thought it was done already. (It needs couple of changes IIRC to the implementation of `wifi_mgmt` in WPA supplicant.

Reply:
Oh brilliant, my bad then. This looks good then. I will re-review one last time, but likely approve